Sources/NIO/ChannelPipeline.swift

```diff
@@ -183,30 +300,55 @@ public final class ChannelPipeline: ChannelInvoker {
         }
 
         let ctx = ChannelHandlerContext(name: name ?? nextName(), handler: handler, pipeline: self)
-        if first {
-            let next = self.head?.next
-            ctx.prev = self.head
-            ctx.next = next
-            head?.next = ctx
-            next?.prev = ctx
-        } else {
-            let prev = self.tail?.prev
-            ctx.prev = prev
-            ctx.next = self.tail
-            tail?.prev = ctx
-            prev?.next = ctx
-        }
+        operation(ctx, relativeContext)
 
         do {
             try ctx.invokeHandlerAdded()
             promise.succeed(result: ())
         } catch let err {
             remove0(ctx: ctx, promise: nil)
-
             promise.fail(error: err)
         }
     }
 
+    /// Synchronously add a single new `ChannelHandlerContext` after one that currently exists in the
+    /// pipeline.
+    ///
+    /// Must be called from within the event loop thread, as it synchronously manipulates the
+    /// `ChannelHandlerContext`s on the `ChannelPipeline`.
+    ///
+    /// - parameters:
+    ///     - new: The `ChannelHandlerContext` to add to the pipeline.
+    ///     - existing: The `ChannelHandlerContext` that `new` will be added after.
+    private func add0(context new: ChannelHandlerContext, after existing: ChannelHandlerContext) {
```

Comment:
I wonder if this method and the one below couldn't have been expressed nicer (without code duplication) with `KeyPath`s. You need one `ReferenceWritableKeyPath` for `next` and one for `prev` and then could can just swap them to do `before` instead of `after`. Am I missing something?

Reply:
You're probably right. These methods are just simple copies of the previous `add` code that split what were previously two branches of an `if` statement into two separate functions.

Reply:
So a quick bit of perf testing suggests that this is not a change we want to make. I perf tested a simple linked list with the duped code vs keypaths, and the key path code is 30x slower. I'm going to chat to the stdlib folks to try to work out if I messed something up, but until I'm sure I don't think the code clarity improvement justifies the 30x slowdown.

Reply:
+1

Reply:
sad but benchmark obviously wins, thanks for doing that @Lukasa